text/router-enhancements.md

```diff
@@ -0,0 +1,211 @@
+
+All segments can be optional, no matter globs, static or dynamic.
+
+### Dynamic segments' constraints
+
+Another feature that can make some usages easier is adding runtime constraints that dynamic segments
+must satisfy for the route to match and that cannot be expressed solely with the micro syntax that
+the router provides.
+
+Consider a weather app that displays the forecast for tomorrow with temperatures, and the user gets
+to choose among a Celsius, Fahrenheit and Kelvin degrees, and preference is stored in URL.
+
+This can be encoded on this string: `/forecast/:scale`. That entry would match `/forecast/celsius`,
+`/forecast/fahrenheit` but also URLs like `/forecast/rankine` which is an unsupported scale.
+
+Currently there is two solutions for this. Either to codify these three routes as static routes
+without dynamic segments, which would leave us with a similar duplication problem as the one
+described for optional segments, or guard against invalid scales in the route and perform a redirect
+via `replaceWith` to the not-found page. However, taking lazy-loading engines into consideration,
+handling constraints manually in the routes requires the app to load and boot the engine in order to
+determine if this URL can be handled, and with route constraints that could be done eagerly by the routes
+and avoid booting the engine at all.
+
+Backend frameworks like Ruby on Rails get a DRY solution to this allowing to receive an executable
+piece of code that the dynamic segment has to validate in order of that route to match.
+
+In the previous example the constraint would be that the scale is one of our white-listed values, and
+any other would cause the route to miss.
+
+
+# Detailed design
+
+
+### Optional segments.
+
+Mimicking rails micro syntax, I propose optional segments to be enclosed between parenthesis. The
+example of the Apple store would be encoded
+
+```js
+Router.map(function() {
+  this.route('product', { path: '(/:region)/:productName' });
+});
+
+
+router.serialize('/mac'); // => routeName='products', params: { productName: 'mac' }
+router.serialize('/uk/mac'); // => routeName='products', params: { productName: 'mac', region: 'uk' }
+```
+
+### Dynamic segment's constraints
+
+Also stealing Ruby on Rails' syntax, the user can provide a constraint using the
+config object passed as second argument.
+
+
+```js
+const allowedScales = ['celsius', 'fahrenheit', 'kelvin'];
+
+function isAllowedScale(scaleName) {
+  return allowedScales.includes(scaleName);
+}
+
+Router.map(function() {
+  this.route('forecast', { path: 'forecast/:scale', constraints: { scale: isAllowedScale } });
+});
+
+
+router.serialize('/forecast/kelvin'); // => routeName='forecast', params: { scale: 'kelvin' }
+router.serialize('/forecast/newton'); // => not recognized route.
+```
+
+Although the generic case of receiving a function that returns a boolean is the most generic one,
+it might worth to add a RegExp support as a convenience.
```

Comment:
Having two invocation patterns puts a branch in a hot loop. I'd prefer only one.